source/extensions/quic_listeners/quiche/envoy_quic_packet_writer.cc

```diff
@@ -0,0 +1,51 @@
+#include "extensions/quic_listeners/quiche/envoy_quic_packet_writer.h"
+
+#pragma GCC diagnostic push
+
+// QUICHE allows unused parameters.
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+// QUICHE uses offsetof().
+#pragma GCC diagnostic ignored "-Winvalid-offsetof"
+
+#include "quiche/quic/core/quic_types.h"
+
+#pragma GCC diagnostic pop
+
+#include "extensions/quic_listeners/quiche/envoy_quic_utils.h"
+#include "common/buffer/buffer_impl.h"
+
+namespace Envoy {
+namespace Quic {
+
+quic::WriteResult EnvoyQuicPacketWriter::WritePacket(const char* buffer, size_t buf_len,
+                                                     const quic::QuicIpAddress& self_ip,
+                                                     const quic::QuicSocketAddress& peer_address,
+                                                     quic::PerPacketOptions* options) {
+  ASSERT(options == nullptr, "Per packet option is not supported yet.");
+  ASSERT(!write_blocked_, "Cannot write while IO handle is blocked.");
+
+  Buffer::BufferFragmentImpl fragment(buffer, buf_len, nullptr);
+  Buffer::OwnedImpl buffer_wrapper;
+  buffer_wrapper.addBufferFragment(fragment);
+  quic::QuicSocketAddress self_address(self_ip, /*port=*/0);
+  Network::Address::InstanceConstSharedPtr local_addr =
```

Comment:
side comment: it seems not great that we are going to be doing this conversion on every packet write, but I'm sure we can optimize that later.

Reply:
As one writer is shared among different connections, it's inevitable to have this conversion on each write. And even within one connection both addresses can change in theory because of connection migration.

If heap allocation is a concern here, maybe we can add a helper function to return a Envoy address on stack. I added a TODO to investigate this later.